clustermesh/clustermesh.go

```diff
@@ -1810,3 +1810,131 @@ func DisableWithHelm(ctx context.Context, k8sClient *k8s.Client, params Paramete
+func (k *K8sClusterMesh) ConnectWithHelm(ctx context.Context) error {
+	v, err := k.client.GetRunningCiliumVersion(ctx, k.params.Namespace)
+	if err != nil {
+		return err
+	}
+	ciliumVer, err := utils.ParseCiliumVersion(v)
+	if err != nil {
+		k.Log("⚠️ Failed to parse Cilium version %s: %w. "+
+			"1.14+ is assumed. Continuing with Helm-mode", v, err)
+	} else {
+		if ciliumVer.LT(versioncheck.MustVersion("1.14.0")) {
+			// Helm-based clustermesh enable is only supported on Cilium v1.14+ due to
+			// a lack of support for autoconfigured certificates (tls.{crt,key}) for cluster
+			// members when running in certgen-based PKI mode
+			return k.Connect(ctx)
+		}
+	}
+
+	remoteCluster, err := k8s.NewClient(k.params.DestinationContext, "")
+	if err != nil {
+		return fmt.Errorf("unable to create Kubernetes client to access remote cluster %q: %w", k.params.DestinationContext, err)
+	}
+
+	aiRemote, err := k.extractAccessInformation(ctx, remoteCluster, k.params.DestinationEndpoints, true, false)
+	if err != nil {
+		k.Log("❌ Unable to retrieve access information of remote cluster %q: %s", remoteCluster.ClusterName(), err)
+		return err
+	}
+
+	if !aiRemote.validate() {
+		return fmt.Errorf("remote cluster has non-unique name (%s) and/or ID (%s)", aiRemote.ClusterName, aiRemote.ClusterID)
+	}
+
+	aiLocal, err := k.extractAccessInformation(ctx, k.client, k.params.SourceEndpoints, true, false)
+	if err != nil {
+		k.Log("❌ Unable to retrieve access information of local cluster %q: %s", k.client.ClusterName(), err)
+		return err
+	}
+
+	if !aiLocal.validate() {
+		return fmt.Errorf("local cluster has the default name (cluster name: %s) and/or ID 0 (cluster ID: %s)",
+			aiLocal.ClusterName, aiLocal.ClusterID)
+	}
+
+	cid, err := strconv.Atoi(aiRemote.ClusterID)
+	if err != nil {
+		return fmt.Errorf("remote cluster has non-numeric cluster ID %s. Only numeric values 1-255 are allowed", aiRemote.ClusterID)
+	}
+	if cid < 1 || cid > 255 {
+		return fmt.Errorf("remote cluster has cluster ID %d out of acceptable range (1-255)", cid)
+	}
+
+	if aiRemote.ClusterName == aiLocal.ClusterName {
+		return fmt.Errorf("remote and local cluster have the same, non-unique name: %s", aiLocal.ClusterName)
+	}
+
+	if aiRemote.ClusterID == aiLocal.ClusterID {
+		return fmt.Errorf("remote and local cluster have the same, non-unique ID: %s", aiLocal.ClusterID)
+	}
+
+	helmValuesLocal := genClusterMeshConfig(aiLocal, aiRemote)
+	// We need a deep copy of these Helm values because `helm.Upgrade` mutates them!
+	helmValuesRemote := genClusterMeshConfig(aiRemote, aiLocal)
+
+	// TODO (ajs): Support hostname-based endpoints via --helm-values override
+	//   Using the Helm value `extraDnsNames` should work as-is.
+	//   This can be useful for LoadBalancer reachability.
+
+	// Enable clustermesh using a Helm Upgrade command
+	upgradeParams := helm.UpgradeParameters{
+		Namespace:   k.params.Namespace,
+		Name:        defaults.HelmReleaseName,
+		Values:      helmValuesLocal,
```

Comment:
Unfortunately, the helm value coalescer is not smart enough to merge arrays; they're always fully replaced.

This means that this will **disconnect** any existing connections, since the `clustermesh.config.clusters` array will be fully overwritten.

You'll have to retrieve the in-use Helm values and mutate them manually, rather than relying on the [automagic helm value coalescing](https://github.com/cilium/cilium-cli/blob/main/vendor/helm.sh/helm/v3/pkg/chartutil/coalesce.go#L36).